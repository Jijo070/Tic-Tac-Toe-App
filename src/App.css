body{
  background: beige;
}

.container{
  text-align: center;
}

.box{

  width: 80px;
  height: 80px;
  border: 1px solid black;
  color: #ff1313;
  font-size: xx-large;
  display: inline-block;
  line-height: 80px;
  cursor:pointer;
  font-weight: 800;
}

img {
  vertical-align: initial;
}

h4 {
  color: #103ab9;
  font-weight: 800;
}

.board{
  display: inline-flex;
  flex-wrap: wrap;
  width: 246px;
}

form{
  margin-bottom: 6px;
}

form label{
  margin-right: 8px;
}

button, input[type="submit" i] {
  border-color: rgb(60, 60, 55);
  border-style: solid;
  border-width: 1px;
  padding: 1px 7px 2px;
  background: lightgrey;
  cursor:pointer;
}

button:hover,
input[type="submit" i]:hover {
  border: 1px solid #0099cc;
  background-color: #00aacc;
  color: #ffffff;
}

button:disabled,
button[disabled]{
  border: 1px solid #999999;
  background-color: #cccccc;
  color: #666666;
}

input[type=checkbox], input[type=radio] {
  margin-left: 4px;
}

.btn{
  text-align: center;
  display: block;
}

.reset{
  display: inline-block;
  margin: auto;
  width: 60px;
  margin-top: 8px;
}

.divider{
    width: 18px;
    height:auto;
    display:inline-block;
}

