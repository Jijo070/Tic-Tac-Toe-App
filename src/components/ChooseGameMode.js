import React, { Component } from 'react';

class ChooseGameMode extends Component{

	handleForm(e){
		e.preventDefault();
		this.props.gamemode(e.target.gamemode.value);
	}

	render(){
		return (
			<form onSubmit={(e)=> this.handleForm(e)}>
				<label>
					Classic
					<input type="radio" name="gamemode" checked="checked" value="Classic"/>
				</label>

				<label>
					Frontenddevlandia
					<input type="radio" name="gamemode" value="Frontenddevlandia"/>
				</label>

				<input type="submit" value="Submit" />
			</form>
		)

	}
}

export default ChooseGameMode;