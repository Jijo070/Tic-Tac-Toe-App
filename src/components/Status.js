import React, { Component } from 'react';

import Player from'./choosePlayer';

class Status extends Component {

	handleSetPlayer(e){
		if (this.props.gamemode) {
			this.props.setPlayer(e)
		}
	}

	renderHtml(){
		if (this.props.winner){
			return (<h5>Winner is {this.props.winner}</h5>)
		} else {
			return this.props.player ? 
	      	<h5>Next player is {this.props.player}</h5> : 
	      	<Player player={(e) => this.handleSetPlayer(e)} />
		}
	}

	render(){

		return (<span>{this.renderHtml()}</span>)
	}
}

export default Status;