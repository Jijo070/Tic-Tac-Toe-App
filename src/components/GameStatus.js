import React, { Component } from 'react';

import ChooseGameMode from'./ChooseGameMode';


class GameStatus extends Component {

	handleSetGameMode(e){
		this.props.setGameMode(e)
	}

	render(){

		return (this.props.gamemode ? 
              <h3>You are playing the {this.props.gamemode} mode</h3> : 
              <ChooseGameMode gamemode={(e) => this.handleSetGameMode(e)} />
        )
	}
}

export default GameStatus;