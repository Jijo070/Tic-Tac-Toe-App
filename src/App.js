import React, { Component } from 'react';

import './App.css';

import Status from'./components/Status';
import GameStatus from'./components/GameStatus';

const connery = require("./images/connery.svg");
const square = require("./images/square.svg");

class App extends Component {


  constructor(props){

    super(props)

    this.state = {

      board : Array(9).fill(null),
      player : null,
      winner : null,
      gamemode : null,
      /* array to store the ndex */
      order_ndex : []
    }

  }

  //Winning conditions
  checkWinner(){

    let winLines =
      [
        ["0", "1", "2"],
        ["3", "4", "5"],
        ["6", "7", "8"],
        ["0", "3", "6"],
        ["1", "4", "7"],
        ["2", "5", "8"], 
        ["0", "4", "8"],
        ["2", "4", "6"]
      ]

    this.checkmatch(winLines)
  }

  //Checks if the current state matches the winning condition
  checkmatch(winLines){
    let board = this.state.board;
    for (let index = 0; index < winLines.length; index++) {
      const [a,b,c]=winLines[index];
      if(board[a] && board[a] === board[b] && board[a] === board[c] ){
        this.setState({
          winner : this.state.player
        })
        this.state.winner = this.state.player;
      }
    }
    if(!this.state.winner && !board.includes(null)){
      this.setState({
          winner : "None"
      })
    }
  }

  handleClick(index){

    if(this.state.player && !this.state.winner){

      let newBoard = this.state.board

      if(this.state.board[index]===null){

        newBoard[index] = this.state.player
        /* push the last index into the array */
        this.state.order_ndex.push(index) 
        this.setState({
          board: newBoard,
          player: this.state.player==="X" ? "O" : "X"
        })

        this.checkWinner()

      }
    }
  } 

  //Set the current state of Player
  setPlayer(player){
    this.setState({player})

  }

  //Set the current state of the Game mode
  setGameMode(gamemode){
    console.log(gamemode)
    this.setState({gamemode})

  }

  //Renders the boxes into DOM
  renderBoxes(){
    const isFrontend = this.state.gamemode === "Frontenddevlandia";
    return this.state.board.map(
      (box, index) => (
      <div 
        className="box" 
        key={index} 
        onClick={() => {
          this.handleClick(index);
        }}
      >
        {box === "X" && isFrontend && <img src={connery} />}
        {box === "O" && isFrontend && <img src={square} />}
        {!isFrontend && box}
      </div>
    ));
  }

  reset(){
    this.setState({
      board : Array(9).fill(null),
      player :  null,
      winner : null,
      setGameMode : null,
      order_ndex : []
    })

  } 

  undo() {
    let ndex = this.state.order_ndex.pop() 
    let newBoard = this.state.board
    let prev = newBoard[ndex]
    newBoard[ndex] = null
    this.setState({
      board: newBoard,
      player: prev
    })
  }


  render() {

    return (
      
      <div className="container">
        <h1>Tic Tac Toe App</h1>

        <GameStatus 
          gamemode ={this.state.gamemode} 
          setGameMode = {(e)=> this.setGameMode(e)}
        />

        <Status
          player={this.state.player} 
          setPlayer={(e) => this.setPlayer(e)}
          gamemode ={this.state.gamemode} 
          winner = {this.state.winner}
        />

        <div className="board">
          
          {this.renderBoxes()}

        </div>
        <div className="btn">
          <button className='reset' onClick = {() => this.reset()}> 
            {" "}
            Reset{" "} 
          </button>
          <div className="divider"/>
          <button 
            className='reset' 
            disabled ={this.state.winner} 
            onClick = {() => this.undo()}
          > 
            {" "}
            Undo{" "} 
          </button>
        </div>
      </div>
    );
  }
}

export default App;
