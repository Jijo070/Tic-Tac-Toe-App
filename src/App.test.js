import ChooseGameMode from './components/ChooseGameMode';
import Player from './components/choosePlayer';
import Status from './components/Status';
import GameStatus from './components/GameStatus'; 
import App from './App';

// Testing for ChooseGameMode component
describe('<ChooseGameMode />',  () => {
	it('renders correctly using snapshot', () => {
	  const editor = render(
	  	<ChooseGameMode />
		);
	expect(editor).toMatchSnapshot();
	});

	it('renders a form', () => {
	  const editor = shallow(<ChooseGameMode />
		);	
	expect(editor.find('form').length).toEqual(1);
	});
});


// Testing for GameStatus component
describe('<GameStatus />',  () => {
	it('renders correctly using snapshot', () => {
	  const wrapper = shallow(
	  	<GameStatus gamemode={"Classic"} setGameMode={() => {}} />
		);
	expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly', () => {
	  const wrapper = render(
	  	<GameStatus gamemode={"Classic"} setGameMode={() => {}} />
		);
	expect(wrapper).toMatchSnapshot();
	});

	it('formats GameStatus correctly ', () => {
	  const wrapper = mount(
	  	<GameStatus gamemode={"Classic"} setGameMode={() => {}} />
		);

	const text = wrapper.find("h3").text();
	expect(text).toEqual("You are playing the Classic mode");
	});

	it('Calls setGameMode correctly on Click', () => {
		const spy =sinon.spy();
	  const wrapper = mount(
	  	<GameStatus gamemode={"Classic"} setGameMode={spy} />
		);
		wrapper.find("h3");
		expect(spy.calledOnce).toBe(false);
	});
});


// Testing for Player component
describe('<Player />',  () => {
	it('renders correctly using snapshot', () => {
	  const editor = render(
	  	<Player />
		);
	expect(editor).toMatchSnapshot();
	});

	it('renders a form', () => {
	  const editor = shallow(<Player />);
	  expect(editor.find('form').length).toEqual(1);
	});
});


// Testing for Status component
describe('<Status />',  () => {
	it('renders correctly using snapshot', () => {
	  const editor = render(
	  	<Status />
		);
	expect(editor).toMatchSnapshot();
	});

	it('renders span', () => {
	  const editor = shallow(<Status />);
	  expect(editor.find('span').length).toEqual(1);
	});

	it('formats Status correctly ', () => {
	  const wrapper = mount(
	  	<Status player={"x"} 
          setPlayer={() => {}}
          gamemode ={"Classic"} 
          winner = {"x"} />
		);

	const text = wrapper.find("h4").text();
	expect(text).toEqual("Game Over! Winner is x");
	});
});



