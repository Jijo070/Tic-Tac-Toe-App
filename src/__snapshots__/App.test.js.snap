// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseGameMode /> renders correctly using snapshot 1`] = `
<form>
  <label>
    Classic
    <input
      name="gamemode"
      type="radio"
      value="Classic"
    />
  </label>
  <label>
    Frontenddevlandia
    <input
      name="gamemode"
      type="radio"
      value="Frontenddevlandia"
    />
  </label>
</form>
`;

exports[`<GameStatus /> renders correctly 1`] = `
<div>
  <h3>
    You are playing the Classic mode
  </h3>
  <form>
    <label>
      Classic
      <input
        name="gamemode"
        type="radio"
        value="Classic"
      />
    </label>
    <label>
      Frontenddevlandia
      <input
        name="gamemode"
        type="radio"
        value="Frontenddevlandia"
      />
    </label>
  </form>
</div>
`;

exports[`<GameStatus /> renders correctly using snapshot 1`] = `
<div>
  <h3>
    You are playing the 
    Classic
     mode
  </h3>
  <ChooseGameMode
    gamemode="Classic"
    setGamemode={[Function]}
  />
</div>
`;

exports[`<Player /> renders correctly using snapshot 1`] = `
<form>
  <label>
    Player X
    <input
      name="player"
      type="radio"
      value="X"
    />
  </label>
  <label>
    Player O
    <input
      name="player"
      type="radio"
      value="O"
    />
  </label>
  <input
    type="submit"
    value="Start"
  />
</form>
`;

exports[`<Status /> renders correctly using snapshot 1`] = `
<span>
  <form>
    <label>
      Player X
      <input
        name="player"
        type="radio"
        value="X"
      />
    </label>
    <label>
      Player O
      <input
        name="player"
        type="radio"
        value="O"
      />
    </label>
    <input
      type="submit"
      value="Start"
    />
  </form>
</span>
`;
